### PR TITLE
refactor: introduce `thiserror-ext` for boxed error wrapper definition

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As explained in #13215. Introducing the `thiserror-ext` proc-macros.

- `Box`: Derive a new type wrapping the `enum` with `Box`. Automatically implement all `From`s and forward error implementation.
- `Construct`: Derive constructors on the new type.

This is a first attempt so I only refactor the `udf::Error` as a demo.

Will release the proc-macros on crates.io after it gets more stable.

----

Derived:
- `Box`
```rust
/// The boxed type of [`ErrorInner`].
#[derive(thiserror_ext::__private::thiserror::Error, Debug)]
#[error(transparent)]
pub struct Error(
    #[from]
    #[backtrace]
    thiserror_ext::__private::ErrorBox<ErrorInner>,
);

impl<E> From<E> for Error
where
    E: Into<ErrorInner>,
{
    fn from(error: E) -> Self {
        Self(thiserror_ext::__private::ErrorBox::new(error.into()))
    }
}
impl Error {
    /// Returns the reference to the inner error.
    pub fn inner(&self) -> &ErrorInner {
        self.0.inner()
    }

    /// Consumes `self` and returns the inner error.
    pub fn into_inner(self) -> ErrorInner {
        self.0.into_inner()
    }
}
```
- `Construct`
```rust
#[automatically_derived]
impl Error {
    /// Constructs a [`ErrorInner::TypeMismatch`] variant.
    pub fn type_mismatch(arg_0: impl Into<String>) -> Self {
        ErrorInner::TypeMismatch { 0: arg_0.into() }.into()
    }

    /// Constructs a [`ErrorInner::Unsupported`] variant.
    pub fn unsupported(arg_0: impl Into<String>) -> Self {
        ErrorInner::Unsupported { 0: arg_0.into() }.into()
    }

    /// Constructs a [`ErrorInner::NoReturned`] variant.
    pub fn no_returned() -> Self {
        ErrorInner::NoReturned {}.into()
    }

    /// Constructs a [`ErrorInner::ServiceError`] variant.
    pub fn service_error(arg_0: impl Into<String>) -> Self {
        ErrorInner::ServiceError { 0: arg_0.into() }.into()
    }
}
```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/udf/src/error.rs [hunk 1]
```diff
@@ -13,21 +13,24 @@
...
 
 /// A specialized `Result` type for UDF operations.
 pub type Result<T, E = Error> = std::result::Result<T, E>;
 
 /// The error type for UDF operations.
-#[derive(thiserror::Error, Debug)]
-pub enum Error {
+#[derive(Error, Debug, Box, Construct)]
+#[thiserror_ext(type = Error)]
+pub enum ErrorInner {
```
Comment (wangrunji0408) [inline src/udf/src/error.rs:25]: nit: can we still define the enum as `Error`? it seems that `ErrorInner` is only used internally in the macro.

File: src/udf/src/external.rs [hunk 2]
```diff
@@ -59,7 +59,7 @@ impl ArrowFlightUdfClient {
         let full_schema = Schema::try_from(info)
             .map_err(|e| FlightError::DecodeError(format!("Error decoding schema: {e}")))?;
         if input_num > full_schema.fields.len() {
-            return Err(Error::ServiceError(format!(
+            return Err(Error::service_error(format!(
```
Comment (wangrunji0408) [inline src/udf/src/external.rs:62]: Perhaps we can keep the name unchanged so that it works fully transparent.

**Review by wangrunji0408 — APPROVED** [review]
LGTM!

**Review by xxchan — COMMENTED** [review]
> This is a first attempt so I only refactor the udf::Error as a demo.

From the PR it's not crystal clear what's the benefit. And related issues are very long. Do you have a TLDR so that anyone can understand it easily? 😄

Specifically, how does this help improve user-facing err msg and/or DX?

**Comment by xxchan** [comment]
> Specifically, how does this help improve user-facing err msg and/or DX?

If this is not clear enough, it might be better to refactor a different error as the demo to show the power.

File: src/udf/src/error.rs (hunk 1 above)
Comment (BugenZhao) [reply src/udf/src/error.rs:25]: They're derive macros so `ErrorInner` is not overwritten and still exists. Developers may still want to match the variants even they get a `Error(Box<ErrorInner>)`, so I also expose `inner` and `into_inner` methods on `Error`.

File: src/udf/src/external.rs (hunk 2 above)
Comment (BugenZhao) [reply src/udf/src/external.rs:62]: `Error` is a struct, so to make it transparent we have to name the constructors in CamelCase. 😕 It seems not a good practice.

**Comment by BugenZhao** [comment]
> Do you have a TLDR so that anyone can understand it easily? 😄

Have you gone through https://github.com/risingwavelabs/risingwave/issues/13215? I tried my best to explain the motivation on this. 🥺

**Comment by BugenZhao** [comment]
Updated the expanded output in the PR body.

**Comment by xxchan** [comment]
> > Do you have a TLDR so that anyone can understand it easily? 😄
> 
> Have you gone through #13215? I tried my best to explain the motivation on this. 🥺

Well, since you are talking about a “demo”, I hope this PR alone can be used to convince and guide others (any dev without enough patience to go through the whole story).🥹

If it’s just “part of” the work, I think it’s OK not too explain further.

File: src/udf/src/external.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/udf/src/external.rs:62]: However, enum variants can be used as constructors by default. 🤔
```rust
enum Error {
    A(i32),
    B(i32, i32),
}

fn main() {
    let f: fn(i32) -> Error = Error::A;
    let f: fn(i32, i32) -> Error = Error::B;
}
```

Comment (BugenZhao) [reply src/udf/src/external.rs:62]: Yeah, but `Error` is a struct instead of enum. 🥵

File: src/udf/src/error.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/udf/src/error.rs:25]: What about doing it in reverse: define the enum as `Error` and generate `ErrorInner` for advanced use.

File: src/udf/src/external.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/udf/src/external.rs:62]: I was always thinking of `Error` as an enum. 🥵 Doesn't matter. The current way LGTM.

**Comment by BugenZhao** [comment]
> If it’s just “part of” the work, I think it’s OK not too explain further.

Yeah. This PR is just for asking for your opinion. Will refactor more in next PRs.

**Pull request merged**